shotover-proxy/src/transforms/redis/cache.rs

```diff
@@ -74,386 +296,493 @@ impl SimpleRedisCache {
+            None => {
+                skipping = true;
+            }
+            Some(relation_elements) => {
+                if skipping {
+                    // we skipped an earlier column so this is an error.
+                    return Err(CacheableState::Err {
+                        reason: "Columns in the middle of the range key were skipped".into(),
+                    });
+                }
+                if relation_elements.len() > 1 {
+                    return Err(CacheableState::Skip {
+                        reason: format!(
+                            "partition key segment {} has more than one relationship",
+                            column_name
+                        ),
+                    });
+                }
+                debug!(
+                    "extending key with segment {} value {}",
+                    column_name, relation_elements[0].value
+                );
 
-            insert_or_update(table_cache_schema, query_values, pk)
+                key.push(b':');
+                key.extend(relation_elements[0].value.to_string().as_bytes());
+            }
         }
-        statement => Err(anyhow!("Cant build query from statement: {}", statement)),
     }
+    Ok(Bytes::from(key))
 }
 
-fn insert_or_update(
+/// build the redis key for the query.
+/// key is cassandra partition key (must be completely specified) prepended to
+/// the cassandra range key (may be partially specified)
+fn build_query_redis_hash_from_value_map(
     table_cache_schema: &TableCacheSchema,
-    query_values: HashMap<String, &Value>,
-    pk: BytesMut,
-) -> Result<RedisFrame> {
-    let mut commands_buffer: Vec<RedisFrame> = vec![
-        RedisFrame::BulkString("ZADD".into()),
-        RedisFrame::BulkString(pk.freeze()),
-    ];
-
-    let clustering = table_cache_schema
-        .range_key
-        .iter()
-        .map(|k| query_values.get(k.as_str()).unwrap())
-        .fold(BytesMut::new(), |mut acc, v| {
-            acc.extend(MessageValue::from(*v).into_str_bytes());
-            acc
-        });
-
-    let values = query_values
-        .iter()
-        .filter_map(|(p, v)| {
-            if table_cache_schema.partition_key.iter().all(|x| x != p)
-                && table_cache_schema.range_key.iter().all(|x| x != p)
-            {
-                Some(MessageValue::from(*v))
-            } else {
-                None
-            }
-        })
-        .collect_vec();
-
-    for v in values {
-        commands_buffer.push(RedisFrame::BulkString(Bytes::from_static(b"0")));
-        let mut value = clustering.clone();
-        if !value.is_empty() {
-            value.put_u8(b':');
-        }
-        value.extend(v.clone().into_str_bytes());
-        commands_buffer.push(RedisFrame::BulkString(value.freeze()));
+    query_values: &BTreeMap<String, Vec<RelationElement>>,
+    select: &Select,
+) -> Result<Bytes, CacheableState> {
+    let mut my_values = query_values.clone();
```

Comment:
we can avoid this clone by doing:
![image](https://user-images.githubusercontent.com/5120858/168539336-5a5064b5-cb26-4b97-9fe2-fb32a54d9595.png)